pennylane/boolean_fn.py

```diff
@@ -77,18 +78,131 @@ class BooleanFn:
 
     """
 
-    def __init__(self, fn):
+    def __init__(self, fn, name=None):
         self.fn = fn
+        self.name = name or self.fn.__name__
         functools.update_wrapper(self, fn)
 
     def __and__(self, other):
-        return BooleanFn(lambda obj: self.fn(obj) and other.fn(obj))
+        return And(self, other)
 
     def __or__(self, other):
-        return BooleanFn(lambda obj: self.fn(obj) or other.fn(obj))
+        return Or(self, other)
+
+    def __xor__(self, other):
+        return Xor(self, other)
 
     def __invert__(self):
-        return BooleanFn(lambda obj: not self.fn(obj))
+        return Not(self)
+
+    def __call__(self, *args, **kwargs):
+        return self.fn(*args, **kwargs)
+
+    def __repr__(self):
+        return f"BooleanFn({self.name})" if not (self.bitwise or self.conditional) else self.name
+
+    @property
+    def bitwise(self):
+        """Determine whether wrapped callable is for a bitwise operation or not."""
+        return bool(getattr(self, "operands", tuple()))
```

Comment:
So this property is for `And`, `Or`, etc.?

Reply:
Yep! This will help us know if the given instance is a result of a bitwise operation.

Reply:
:thinking: any reason why not to use instance checks for this?

Reply:
We could do instance checks as well. I find this to be more generalized as it makes this property robust to any bitwise class that is added next or is user-made.